components/suggest/src/suggest.udl

```diff
@@ -125,6 +133,9 @@ dictionary SuggestIngestionConstraints {
     // there is a schema upgrade that causes the database to be thrown away, then the
     // `empty_only=true` ingestion that runs on startup will repopulate it.
     boolean empty_only = false;
+    // If `providers` contains `Exposure`, this must be set to an exposure
+    // suggestion type to ingest.
+    string? exposure_suggestion_type;
```

Comment:
I don't love the way this field is coupled with the exposure type (this also goes for `SuggestionQuery`).  I'd rather one of these solutions:

 - Changing `SuggestionProvider` from a simple enum to one with fields.  The only variant that would have a field is `SuggestionProvider::Exposure` and it would get a `suggestion_type` field.
 - Creating a separate provider for each exposure type (`SuggestionProvider::ExposureFakespot`, `SuggestionProvider::ExposureAmp`, etc).

I think this would work better with the newly merged code.  Also, it would let consumers query for multiple exposure types at once.

Reply:
Yeah, I don't like it much either. The second solution won't work because the whole point of this is to not require any code changes to run new exposure experiments. I like your first solution so I'll try that.

Reply:
My latest commit converts `SuggestionProvider` from a unit-only enum to a non-unit-only enum. There weren't as many resulting changes as I thought there would need to be, which I guess means it's not a very consequential change. I do think it makes it a little harder to understand what "provider" means. To me, each use of a `SuggestionProvider` is now more like an "instance" than a simple static const value, but that might just be my Rust newbie-ness showing.

Reply:
IMO this is the main thing we need to get right.  I don't think this is your "Rust newbie-ness", the current patch changes the meaning of "provider" and the result is confusing.  For example, the `suggestions` table has a provider field which is an int, which is no longer really correct.  Maybe we could make this work by treating providers like records by having both a `SuggestionProvider` and `SuggestionProviderType` types, but I'm not sure I like that.

Would it be possible to implement this by adding a `SuggestionProvider::Exposure` variant without the `suggestion_type` field?  The suggestion type is already stored an a per-suggestion basis so this should be doable:
  - Whenever a query includes `SuggestionProvider::Exposure` we'd return all exposure suggestions that matched, and the client could filter them based on the `Suggestion::Exposure::type` field and the enabled experiment features.   
 - When ingesting, we'd ingest all exposure suggestions together if `SuggestionProvider::Exposure` was specified.  That might be slightly wasteful, but would it be too bad?

Reply:
I don't really like the idea of ingesting and fetching all exposure types. With translations suggestions for example, we're going to have en-US, fr, and de versions. It's not a good design to have any user enrolled in a translations experiment download all localizations.

Tbh I was thinking simply renaming `Provider` to something like `ProviderConfig`, `ProviderOptions`, or `ProviderInstance` might make it feel right.

The other thought I had was introducing a separate parallel `ProviderOptions` enum, and then you would pass them on query and ingest. But I don't know how to force consumers at compile-time to pass correct options given the providers they also passed. It would have to be a runtime error if the options don't match the providers, or we could just silently ignore it I guess.

Reply:
> The other thought I had was introducing a separate parallel `ProviderOptions` enum, and then you would pass them on query and ingest. 

This would be a generalization of my first approach in this PR. Instead of exposure-specific fields in `SuggestionQuery` and `SuggestIngestionConstraints`, we'd have a generalized `provider_options` field -- a `sequence` I guess. Is there a map type in udl/uniffi? If the options could be a map from `SuggestionProvider` (as a plain `enum` variant) to `SuggestionProviderOptions`, that would be nice. But I don't know how to statically force consumers to use the `SuggestionProviderOptions` that matches the `SuggestionProvider`.

I do wonder whether exposure suggestions will always be the one special case though. If so, the simplest thing is my first approach IMO. Maybe we could go with that for now and cross the generalized bridge if/when we get to it?

Reply:
> The other thought I had was introducing a separate parallel ProviderOptions enum, and then you would pass them on query and ingest.

I like this idea.

It's possible to create a map in UDL using `record<KeyType, ValueType>`.  "record" is kind of counter-intuitive here, but it's inherited form IDL.  The keys used to have to be string, but I think you can now use enum values nowadays.

However, I think I prefer using a struct and having each suggestion type ignore any fields that didn't apply to them.  Initially, it could just have a single  `exposure_suggestion_types` field.  Later on, if we added a `locale` field, that couple apply to exposure suggestions, AMP, and any other suggestion type that had localization support.

Reply:
That sounds good to me. I've come around to thinking we should go back to the first approach or something like it to keep things simple given the use case and not over-engineer or over-generalize. A simple provider options struct seems like a good compromise between the first approach and a more generalized approach. We can always re-evaluate if more providers end up needing ingest- or query-time options.

Reply:
I called it `SuggestionProviderConstraints` since it's analogous to ingest constraints and I didn't want to cause confusion with the existing `SuggestProviderConfig`, which is something else entirely. Let me know what you think.